Reduce allocation in MethodTypes.isTrivial

diesalbla opened the pull request:
Fixes scala/scala-dev#492 

This PR addresses [Issue 492](https://github.com/scala/scala-dev/issues/492). It aims at reducing the allocation of Cons objects in the execution of the `isTrivial` method. Along the way, it also tries to reduce the number of `ContainsCollector` objects that are getting allocated.

diesalbla committed:
Solve Issue 492, "MethodTypes.isTrivial is allocation heavy"

This commit addresses Issue 492 at ScalaDev.
It aims at reducing the allocation of Cons objects in the execution of
the `isTrivial` method. In particular:

- We use an array, evaluated on demand, to contain the `_.tpe` of each
  parameter in `params`.
- We also add an array for the ContainsCollector objects, which is
  also one for each parameter. Before, the method would create one
  for each type, which meant N^2 creations in the worst case.
- We turn the auxiliary methods for `isTrivial` from recursive into
  imperative loops, using when possible indexes over those arrays.

retronym commented:
We don’t want to spend any fields here. This is only called once per
MethodType, and the final bool is cached.

diesalbla committed:
Move fields into local variables.

The method `areTrivialParams` is assumed to be used at most once
per MethodType object, so there is no need to keep the arrays as
fields. Instead, we can turn them into local method variables.

- Since the main loop goes parameter by parameter, we need not
  to keep the ContainsCollector objects, only the types.

- We unroll the main loop, so that we first check the `.isTrivial`
  method of each type, and only afterwards check with anyContains.

diesalbla commented:
@retronym I have modified accordingly to your last comment. One array is no longer needed, and the other one is just a local variable of the method.

retronym commented:
@diesalbla Could you please review and accept [our CLA](https://www.lightbend.com/contribute/cla/scala)?

retronym commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -2573,20 +2573,61 @@ trait Types
+            paramsTpes(ix) = pps.head.tpe
+            res = paramsTpes(ix).isTrivial
+            pps = pps.tail
+            ix = ix + 1
```
Let's just use `ix += 1` here and below, which is the style we typically use elsewhere.

retronym commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -2573,20 +2573,61 @@ trait Types
+    private def areTrivialParams: Boolean =
+      if (params.isEmpty) true else {
+        val len = params.length
+        val paramsTpes: Array[Type] = new Array[Type](len)
```
This was the only caller to `private def typesContain` so it should be deleted.

retronym committed:
Minor style changes and remove dead code

retronym commented:
I've pushed those minor changes to this PR and will merge once CI runs over the new commit.

Thanks for your contribution!
